eeps/eep-0072.md

```diff
@@ -0,0 +1,145 @@
+    Author: Jesse Gumm <ja(at)gumm(dot)io>
+    Status: Draft
+    Type: Standards Track
+    Created: 08-Oct-2024
+    Post-History:
+****
+EEP 72: Reserved words and Variables as record names, and enhancement to definition syntax
+----
+
+Abstract
+========
+
+This EEP loosens some of the restrictions around record names such no longer is
+it necessary to quote record names when those records are named with reserved
+words (`#if` vs `#'if'`) or words with capitalized first characters (`#Hello`
+vs `#'Hello'`).
+
+This EEP also proposes to add a new record-like syntax to the record
+definitions (also adopting the above syntactical changes), such that the
+following record definitions would all be valid and identical:
+
+```erlang
+-record('div', {a :: integer(), b :: integer()}).
+-record #'div'{a :: integer(), b :: integer()}.
+-record #div{a :: integer(), b :: integer()}.
+```
+
+Motivation
```

Comment:
The motivation section covers the reason behind the usage of unquoted "keywords" atoms, but not the eventual benefits from the introduction of the new record syntax. I can see how the latter could be more opinionated than the former. Maybe we should limit this EEP to the usage of unquoted atoms only and, if there's really a benefit from multiple syntaxes for record definitions, have a separate EEP/motivation?

Reply:
The motivation I'd say is to be able to use the same quoting for record definition as for usage.  Otherwise one would have to know which atoms that has to be quoted in the record definition, which is awkward and error prone, or always quote the atom in the record definition, which is ugly.

But I agree that this should be in the Motivation chapter.

There should also be a motivation for why we want to allow variable names unquoted, and that is for applications such as ASN.1 or Corba that have such name conventions and that now have to quote the record names just like for the `'div'` example.

Separating into two different EEP:s I think would be unfortunate, because we then might get a period of asymmetry when one cannot use the same format in record definition and usage.

That there are so many variants of how to define a record is largely due to the syntax rules for attributes and atom quoting in Erlang.

Maybe we should instead write that the old and the new syntaxes are:

```erlang
-record('div', {a :: integer(), b :: integer()}).
-record #div{a :: integer(), b :: integer()}.
```

and then elaborate that due to parentheses being optional in attributes, and due to that atoms may be quoted even when not needed, these happen to be equivalent, just mentioned for completeness:

```erlang
-record 'div', {a :: integer(), b :: integer()}.
-record(#div{a :: integer(), b :: integer()}).
-record #'div'{a :: integer(), b :: integer()}.
-record(#'div'{a :: integer(), b :: integer()}).
```